cnqr/_src/convex.py

```diff
@@ -0,0 +1,207 @@
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_data
+      f_v.append(
+          self.context_dense_layer(
+              output_dim,
+              name=f'f_v_dense_{i}',
+              kernel_init=self.init_fn(self.init_std),
+              bias_init=nn.initializers.constant(0.0),
+              use_bias=True,
+          )
+      )
+    self.f_v = f_v
+
+    f_w = []  # dense layers (with bias), init: weights -> approx(0), bias -> 0
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
+      f_w.append(
+          self.context_dense_layer(
+              output_dim,
+              name=f'f_w_dense_{i}',
+              kernel_init=self.init_fn(self.init_std),
+              bias_init=nn.initializers.constant(0.0),
+              use_bias=False,
+          )
+      )
+    self.f_w = f_w
+
+    # ===== Convex network =====
+    w_U = []  # positive dense layers (no bias), init: weights -> ones / d1
+    normalization = 1
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
+      w_U.append(
+          PositiveDense(
+              output_dim,
+              name=f'w_U_pdense_{i}',
+              kernel_init=initializers.constant(rescale(1.0 / normalization)),
+              use_bias=False,
+          )
+      )
+      if i < self.num_hidden:
+        normalization = self.dim_hidden[i]
+    self.w_U = w_U
+
+    w_V = []  # dense layers (no bias), init: weights -> approx(0)
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
+      w_V.append(
+          nn.Dense(
+              output_dim,
+              name=f'w_V_dense_{i}',
+              kernel_init=self.init_fn(self.init_std),
+              use_bias=False,
+          )
+      )
+    self.w_V = w_V
+
+  def _compute_identity_map(self, input_dim):
+    A = jnp.eye(input_dim).reshape((1, input_dim, input_dim))
+    b = jnp.zeros((1, input_dim))
+    return A, b
+
+  @staticmethod
+  def _identity_initializer(key: KeyArray,
+                            shape: core.Shape,
+                            dtype: DTypeLikeInexact = jnp.float_) -> Array:
+    return jnp.eye(N=shape[0], M=shape[1], dtype=dtype)
+
+  @nn.compact
+  def __call__(self, x, y):
+    """
+    Args:
+        x: Context
+        y: Observation from source distribution
+    """
+    # TODO: put everything in the call function (delete setup)
+    x = self.phi(x)
+    z = self.q(y) * jnp.ones(shape=x.shape)
+    for i in range(self.num_hidden + 2):
+      # context network
+      x = self.act_fn_x(self.f_x[i](x))
+      # context modulators
+      u = self.f_u[i](x)
+      v = self.f_v[i](x)
+      w = self.f_w[i](x)
+      # convex network
+      z = self.act_fn_z(self.w_U[i](z * nn.softplus(u)) + self.w_V[i](y * v) + w)
```

Comment:
Add the variable `h_k` to split computation of `self.w_U[i](z * nn.softplus(u)) + self.w_V[i](y * v) + w` and  `self.act_fn_z`.